rcl_action/test/rcl_action/test_graph.cpp

```diff
@@ -0,0 +1,548 @@
+    &this->node, &this->allocator, "_test_this_Isnot_a_valid_name", "", nullptr);
+  EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+  ret = rcl_action_get_server_names_and_types_by_node(
+    &this->node, &this->allocator, this->test_graph_old_node_name, "", nullptr);
+  EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+  // Invalid names and types
+  ret = rcl_action_get_server_names_and_types_by_node(
+    &this->node, &this->allocator, this->test_graph_node_name, "", nullptr);
+  EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+  // Valid call
+  ret = rcl_action_get_server_names_and_types_by_node(
+    &this->node, &this->allocator, this->test_graph_node_name, "", &nat);
+  EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+  ret = rcl_names_and_types_fini(&nat);
+  EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+}
+
+TEST_F(
+  CLASSNAME(TestActionGraphFixture, RMW_IMPLEMENTATION),
+  test_action_get_names_and_types)
+{
+  rcl_ret_t ret;
+  rcl_node_t zero_node = rcl_get_zero_initialized_node();
+  rcl_names_and_types_t nat = rcl_get_zero_initialized_names_and_types();
+
+  // Invalid node
+  ret = rcl_action_get_names_and_types(nullptr, &this->allocator, &nat);
+  EXPECT_EQ(RCL_RET_NODE_INVALID, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+  ret = rcl_action_get_names_and_types(&zero_node, &this->allocator, &nat);
+  EXPECT_EQ(RCL_RET_NODE_INVALID, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+  ret = rcl_action_get_names_and_types(&this->old_node, &this->allocator, &nat);
+  EXPECT_EQ(RCL_RET_NODE_INVALID, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+  // Invalid allocator
+  ret = rcl_action_get_names_and_types(&this->node, nullptr, &nat);
+  EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+  // Invalid names and types
+  ret = rcl_action_get_names_and_types(&this->node, &this->allocator, nullptr);
+  EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+  // Valid call
+  ret = rcl_action_get_names_and_types(&this->node, &this->allocator, &nat);
+  EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+  ret = rcl_names_and_types_fini(&nat);
+  EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+}
+
+/**
+ * Type define a get actions function.
+ */
+typedef std::function<rcl_ret_t(const rcl_node_t *,
+    rcl_names_and_types_t *)> GetActionsFunc;
+
+/**
+ * Extend the TestActionGraphFixture with a multi-node fixture for node discovery and node-graph
+ * perspective.
+ */
+class TestActionGraphMultiNodeFixture : public CLASSNAME(TestActionGraphFixture, RMW_IMPLEMENTATION)
+{
+public:
+  const char * remote_node_name = "remote_graph_node";
+  const char * action_name = "/test_action_info_functions__";
```

Comment:
http://design.ros2.org/articles/topic_and_service_names.html forbids `__`

> * must not contain any number of repeated underscores (`_`)

I'm surprised the action topics and services are created successfully. This might be worth opening a bug for.

Reply:
I didn't realize, I copied this from what's done in the `rcl` graph tests:

https://github.com/ros2/rcl/blob/d810f12e94ae9498f34ebcf77951b544d21aa10f/rcl/test/rcl/test_graph.cpp#L737